week-1-ts-ecosystem/w01e03-object-oriented-structure/src/feature/cart/components/cart.ts

```diff
@@ -0,0 +1,76 @@
+import { Product } from './product';
+
+interface Cart {
+  products: Product[];
+  cartType: 'Auction' | 'Buy Now' | 'Free';
+}
+
+class Cart {
+  constructor(cartType: 'Auction' | 'Buy Now' | 'Free') {
+    this.products = [];
+    this.cartType = cartType;
+  }
+
+  private isProductExist(id: number) {
+    return this.products.some((product) => product.id === id);
+  }
+
+  getAllProducts() {
+    return this.products;
+  }
+
+  addProduct(id: number, name: string, price: number, amount: number, type: 'Auction' | 'Buy Now' | 'Free') {
+    if (this.isProductExist(id)) {
+      return console.log(`You can not add products with this same ID`);
+    }
+    if (this.cartType === type) {
+      this.products.push(new Product(id, name, price, amount, type));
+    } else {
+      return console.log(`You can not add products with type ${type} to cart with type ${this.cartType}`);
+    }
+  }
+
+  getProductById(id: number) {
+    if (this.isProductExist(id)) {
+      return this.products.filter((product) => product.id === id);
+    } else {
+      return console.log(`There is no product with such ID.`);
+    }
+  }
+
+  updateProductById(id: number, productToUpdate: Product) {
+    if (id !== productToUpdate.id) return console.log('UPDATE: ID need to be this same.');
+    if (this.cartType === productToUpdate.type) {
+      this.products = this.products.map((product) => (product.id === id ? productToUpdate : product));
+    } else {
+      return console.log(
+        `You can not update products with type ${productToUpdate.type} to cart with type ${this.cartType}`
+      );
+    }
+  }
+
+  deleteProductById(id: number) {
+    if (this.isProductExist(id)) {
+      this.products = this.products.filter((product) => product.id !== id);
+    } else {
+      console.log(`DELETE: There is no product with this ID.`);
```

Comment:
podobnie jak wyżej z obsługą błędów oraz `if...else` może warto odwrócić warunek zobacz:

```typescript
if (!this.isProductExist(id)) {
  throw new Error(`DELETE: There is no product with this ID: ${id}.`)   
}
this.products = this.products.filter((product) => product.id !== id);     
```

if (this.isProductExist(id)) {
      this.products = this.products.filter((product) => product.id !== id);
    } else {
      console.log(`DELETE: There is no product with this ID.`);